### ISSUE TITLE
View Functions Error: attached_deposit prohibited

### ISSUE THREAD

**Issue by ewilz** [near/near-evm#19]
Even though we don't seem to explicitly call attached_deposit from `view_call_contract`:

 Error: invalid hexidecimal string (arg="value", value="0xError: Querying call/near_evm/view_call_contract failed: wasm execution failed with error: FunctionCallError(HostError(ProhibitedInView { method_name: \"attached_deposit\" })).\n{\n  \"error\": \"wasm execution failed with error: FunctionCallError(HostError(ProhibitedInView { method_name: \\\"attached_deposit\\\" }))\",\n  \"logs\": [],\n  \"block_height\": 209,\n  \"block_hash\": \"BTiAQgoFfFhgiSJ9PG3ZbBRH44ydhBUoBKp5yqm9AB2b\"\n}")

@ilblackdragon @bowenwang1996

**Comment by bowenwang1996** [comment]
Can you give a link to the code that caused this failure? It potentially calls some other function that tries to access `attached_deposit`

**Comment by barbaraliau** [comment]
@bowenwang1996 I just opened this issue in `near-web3-provider` and then saw @emaG3m had the same issue.

https://github.com/nearprotocol/near-web3-provider/issues/19

This is the code that is called in the tests:

https://github.com/nearprotocol/near-web3-provider/blob/master/src/index.js#L41
https://github.com/nearprotocol/near-web3-provider/blob/master/src/index.js#L705

```js
    async _viewEvmContract(method, methodArgs) {
        try {
            const result = await this.account.viewFunction(
                this.evm_contract,
                method,
                methodArgs
            );
            return result;
        } catch (e) {
            return e;
        }
    }

---------------------------
    /**
     * Executes a new message call immediately without creating a
     * transaction on the block chain
     * @param {Object} txCallObj transaction call object
     * @property {String} to the address the tx is directed to
     * @property {String} from (optional) the address the tx is sent from
     * @property {Quantity} value (optional) integer of the value sent
     * with this tx
     * @property {String} data (optional) hash of the method signature
     * and encoded parameters
     * @returns {String} the return value of the executed contract
     */
    async routeEthCall([txCallObj]) {
        try {
            const { to, from, value, data } = txCallObj;
            const sender = from
                ? from
                : utils.nearAccountToEvmAddress(this.accountId);

            const val = value
                ? new BN(utils.remove0x(value), 16)
                : new BN(0);
            const result = await this._viewEvmContract(
                'view_call_contract',
                {
                    contract_address: utils.remove0x(to),
                    encoded_input: utils.remove0x(data),
                    sender: utils.remove0x(sender),
                    value: val.toString()
                });
            return '0x' + result;
        } catch (e) {
            return e;
        }
    }
```

**Comment by prestwich** [comment]
It seems like something has changed in the env behavior, as that code path does not directly call `attached_deposit`

**Comment by bowenwang1996** [comment]
Not clear to me what happened. @evgenykuzyakov is there some recent change in near-sdk-rs that uses `attached_deposit`?

**Comment by emaG3m** [comment]
@bowenwang1996 @evgenykuzyakov I tried debugging this a bit, and we expect our function `view_call_contract` to be called for view functions, but I'm noticing I don't seem to even reach this function by the time the error `ProhibitedInView { method_name: "attached_deposit" }))` is returned. 

This function's implementation uses the `[#near_bindgen]` macro.

**Comment by ilblackdragon** [comment]
The reason is that `view_call_function` takes self as mutable reference. View calls need to be immutable references `&self` but this needs some major refactoring :(

**Comment by bowenwang1996** [comment]
> View calls need to be immutable references &self

When did we start requiring that?

**Comment by prestwich** [comment]
Transcribing info from Telegram (thanks @nearmax) 

First implement `pub fn view_call_contract` in a separate `impl EvmContract` block without `#[near_bindgen_macro]`. Then declare the following outside:
```
#[cfg(target_arch = "wasm32")]
#[no_mangle]
pub extern "C" fn view_call_contract() {
    near_sdk::env::setup_panic_hook();
    near_sdk::env::set_blockchain_interface(Box::new(near_blockchain::NearBlockchain {}));

    #[derive(serde::Deserialize, serde::Serialize)]
    struct Input {
        contract_address: String,
        encoded_input: String,
        sender: String,
        value: Balance,
    }
    let Input { contract_address, encoded_input,sender, value }: Input = serde_json::from_slice(
         &near_sdk::env::input().expect("Expected input since method has arguments.")
    ).expect("Failed to deserialize input from JSON.");

    let mut contract: EvmContract = near_sdk::env::state_read().unwrap_or_default();
    let result = contract.method(contract_address, encoded_input,sender, value, );
    let result = serde_json::to_vec(&result).expect("Failed to serialize the return value using JSON.");
    near_sdk::env::value_return(&result);
 }
```

**Comment by prestwich** [comment]
@barbaraliau can you confirm that this is fixed in latest master?  (by #22)

**Comment by barbaraliau** [comment]
@prestwich i've updated the wasm code on the provider and tests are passing. 👍

**Comment by prestwich** [comment]
closed by #22
